[Feature-10683][Task Plugin] Add Java Task Plugin.

106umao opened the pull request:
<!--Thanks very much for contributing to Apache DolphinScheduler. Please review https://dolphinscheduler.apache.org/en-us/community/development/pull-request.html before opening a pull request.-->

This feature close #10683 

## Purpose of the pull request

<!--(For example: This pull request adds checkstyle plugin).-->

## Brief change log

<!--*(for example:)*
  - *Add maven-checkstyle-plugin to root pom.xml*
-->
## Verify this pull request

<!--*(Please pick either of the following options)*-->

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

<!--*(example:)*
  - *Added dolphinscheduler-dao tests for end-to-end.*
  - *Added CronUtilsTest to verify the change.*
  - *Manually verified the change by testing locally.* -->

106umao committed:
[Feature] init task plugin for java type

106umao committed:
Finish The front end work have finished almost

106umao committed:
[Feature] add unit test for java type task plugin.

106umao committed:
update optimize the unit test.

106umao committed:
update move all uml meta to folder uml.

106umao committed:
update improve front end content.

106umao committed:
update add java task plugin maven info to project task all.

106umao committed:
test add some test change to optimize develop this branch.

106umao committed:
feature add java task plugin.

106umao committed:
fix mainjar bug, resource path bug.

106umao committed:
update add isModulePath , remove java version.

106umao committed:
update add module path support.

106umao committed:
add add doc with en and ch.

106umao committed:
update remove the jmod configuration.
update remove uml diagram

update adjust doc-en's content

update restore common configuration.

update restore AbstractYarnTask.java

update delete servlet dependency in javatask pom.

update adjust java task .

update restore ui's service.ts

106umao committed:
update add license for java task pom file.

106umao committed:
fix the ts2322 in format-data.ts

106umao committed:
fix the interrupt exception.

106umao committed:
update adjust code style

106umao committed:
fix the typescript error TS2352.

106umao committed:
fix the java task param map bug

zhongjiajie commented on docs/docs/en/guide/task/java.md:
```diff
@@ -0,0 +1,67 @@
+# Task Parameters
+
+- Node Name: the name of the set task. The node name in a workflow definition is unique.
+
```
please remove the redundant blank line here

106umao replied:
thx, it is done

zhongjiajie commented on docs/docs/en/guide/task/java.md:
```diff
@@ -0,0 +1,67 @@
+The main configuration parameters are as follows:
+
+- Run Type
+
```
redundant blank line too

106umao replied:
thx, it is done

zhongjiajie commented on docs/docs/zh/guide/task/java.md:
```diff
@@ -0,0 +1,50 @@
+- 虚拟机参数
+- 脚本文件
+
+![java_task](../../../../img/tasks/demo/java_task01.png)
```
both English and Chinese documents should use the English version screenshot

106umao replied:
thx, it is done

zhongjiajie commented on docs/docs/en/guide/task/java.md:
```diff
@@ -0,0 +1,67 @@
+
+![java_task](../../../../img/tasks/demo/java_task02.png)
+
+## Notice
```
```suggestion
## Note
```

106umao replied:
thx, it is done

Amy0104 commented on dolphinscheduler-ui/src/locales/en_US/project.ts:
```diff
@@ -126,7 +126,6 @@ export default {
     switch_version: 'Switch To This Version',
     confirm_switch_version: 'Confirm Switch To This Version?',
     current_version: 'Current Version',
-    run_type: 'Run Type',
```
The 'run_type' can not be removed, because it is used in workflow instance column.

106umao replied:
> The 'run_type' can not be removed, because it is used in workflow instance column.

thank you. maybe I deleted it by mistake.

Amy0104 commented on dolphinscheduler-ui/src/locales/zh_CN/project.ts:
```diff
@@ -284,6 +284,10 @@ export default {
+    is_module_path: '模块路劲',
+    run_type: '运行类型',
+    jvm_args: '虚拟机参数',
+    jvm_args_tips: '请输入虚拟机参数',
```
It is better to keep these keys the same position as en_US.

106umao replied:
thx, it is done

kezhenxu94 commented on dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/model/TaskResponse.java:
```diff
@@ -123,4 +123,18 @@ public TaskRunStatus getStatus() {
+                ", exitStatusCode=" + exitStatusCode +
+                ", status=" + status +
+                '}';
+    }
```
Please use `@ToString` annotation (or even `@Data` annotation and remove the `getter/setter`) from lombok

106umao replied:
thx, it is done

kezhenxu94 commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaConstants.java:
```diff
@@ -0,0 +1,46 @@
+    public static final String RUN_TYPE_JAVA = "JAVA";
+    public static final String RUN_TYPE_JAR = "JAR";
+
+//    to be extended.
+//    public static final String RUN_TYPE_JSHELL = "JSHELL";
+//    public static final String RUN_TYPE_INJVM = "INJVM";
+
+
+
```
Do put these commented codes here if you don't plan to implement in this PR, create an issue to track the todo tasks

106umao replied:
thx, it is done

kezhenxu94 commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaParameters.java:
```diff
@@ -0,0 +1,114 @@
+     */
+    private List<ResourceInfo> resourceList;
+
+    public String getRawScript() {
```
Use `@Getter/@Setter` annotations from lombok

106umao replied:
thx, it is done

Amy0104 commented on dolphinscheduler-ui/src/views/projects/task/components/node/fields/use-java-task-main-jar.ts:
```diff
@@ -0,0 +1,80 @@
+    },
+    options: mainJarOptions
+  }
+}
```
It seems that this hook is not used at all.

106umao replied:
thx, changes have been made to make Java tasks better code-isolated from other tasks

Amy0104 replied:
It already has a useMainJar hook, so it don't need to create a new one.

Amy0104 commented on dolphinscheduler-ui/src/views/projects/task/components/node/format-data.ts:
```diff
@@ -35,6 +35,19 @@ export function formatParams(data: INodeData): {
+      if (data.mainJar) {
+        taskParams.mainJar = { id: data.mainJar }
+      }
+    }
```
it is better to change the two conditions into one parallel condition.

106umao replied:
thx, it is done

Amy0104 commented on dolphinscheduler-ui/src/views/projects/task/components/node/fields/use-java.ts:
```diff
@@ -0,0 +1,98 @@
+            //     model.mainJar = null
+            //     model.rawScript=''
+            // }
+          }
```
It is better to remove the comment code.

106umao replied:
thx,it is done

106umao committed:
optimize the code style.

106umao committed:
chenged the java task main jar reference.

106umao committed:
remove the unused img.

106umao committed:
fix typo of doc.

106umao committed:
optimize the code style.

106umao committed:
fix the problem of checkstyle:checkstyle-aggregate

SbloodyS commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/test/org/apache/dolphinscheduler/plugin/task/java/JavaTaskTest.java:
```diff
@@ -0,0 +1,161 @@
+        property.setDirect(IN);
+        property.setType(VARCHAR);
+        javaParameters.setLocalParams(localParams);
+//        javaParameters.setVarPool("");
```
We should remove meaningless code.

106umao replied:
thx, it is done .

106umao committed:
optimize the code style.

106umao committed:
adjust the exception type.

pinkhello commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaTask.java:
```diff
@@ -0,0 +1,343 @@
+
+    public String getPublicClassName(String sourceCode) {
+        String pattern = "(.*\\s+public\\s+class\\s+)([a-zA-Z_]+[//w_]*)([.\\s\\S]*)";
+        Pattern compile = Pattern.compile(pattern);
```
I think should cache regex pattern to improve performance, its performance matters.

106umao replied:
good idea, have done.

pinkhello commented:
SonarCloud CI failed

pinkhello commented:
OWASP Dependency Check  CI Failed

zhuangchong commented on dolphinscheduler-ui/src/views/projects/task/components/node/fields/use-main-jar.ts:
```diff
@@ -26,7 +26,7 @@ export function useMainJar(model: { [field: string]: any }): IJsonItem {
   const taskStore = useTaskNodeStore()
 
-  const mainJarSpan = computed(() => (model.programType === 'SQL' ? 0 : 24))
+  const mainJarSpan = computed(() => (model.programType === 'SQL' ? 0 : 24));
```
No need to modify here.

106umao replied:
it is done

zhuangchong reviewed (changes requested):
You need to add the JAVA task documentation link in the docsdev.js file.

https://github.com/apache/dolphinscheduler/blob/dev/docs/configs/docsdev.js

106umao committed:
update cached the regex, add link into docsdev.js for java task

106umao committed:
update remvoe ; in user-main-jar.ts

106umao commented:
> You need to add the JAVA task documentation link in the docsdev.js file.
> 
> https://github.com/apache/dolphinscheduler/blob/dev/docs/configs/docsdev.js

thx, have done

106umao committed:
fix typo

resolve the conflict, modify @Data to e.g. @Setter, @Getter

modify the construction way for java command path.

fix typo

zhuangchong reviewed (dismissed):
LGTM.

zhuangchong commented:
@Amy0104 Please help code review.

zhuangchong commented:
@zhongjiajie @SbloodyS @Amy0104 Please help code review. thanks.

SbloodyS commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaTask.java:
```diff
@@ -0,0 +1,348 @@
+    }
+
+    private String getJavaHomeBinAbsolutePath() {
+        String javaHomeAbsolutePath = System.getenv(JavaConstants.JAVA_HOME);
```
It's better to use `${JAVA_HOME}` instead of this since this would invalidate the setting of environmental management. Just like we do in the python task plugin.

At present, we write the configuration of environment management into the shell script and execute it through `source` to make the environment variables take effect.
https://github.com/apache/dolphinscheduler/blob/25ded1e81b59a216275e6bbe26e0eb3e6f400f78/dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/ShellCommandExecutor.java#L103-L114

106umao replied:
Hi @SbloodyS 
thank you, I have modified for it now , PTAL.

EricGao888 commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaParameters.java:
```diff
@@ -0,0 +1,64 @@
+
+import lombok.Data;
+
+@Data
```
Hi~ I recommend replacing `@Data` with `@Getter`, `@Setter` and `@ToString` because `@Data` will generate some code we may not need and cause the test coverage to decrease. Thx : )
see: https://projectlombok.org/features/Data and https://stackoverflow.com/questions/45569085/sonarqube-bad-coverage-because-of-lombok-data

EricGao888 replied:
> SonarCloud Quality Gate failed.    [![Quality Gate failed](https://camo.githubusercontent.com/4ea51c1f64ee3746f631653a02ab678ca6a3efb5f5cb474402faed2e3dcf90b5/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f5175616c6974794761746542616467652f6661696c65642d313670782e706e67)](https://sonarcloud.io/dashboard?id=apache-dolphinscheduler&pullRequest=10689)
> 
> [![Bug](https://camo.githubusercontent.com/4c6102327f5a954f9c8acaf2e2714183157a9e41717b371b2cd585cf25057310/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f6275672d313670782e706e67)](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=10689&resolved=false&types=BUG) [![A](https://camo.githubusercontent.com/1cba125a897d7fa47033a3b3b2be2bbee680d34d4f004a215564659b853fb201/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f412d313670782e706e67)](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=10689&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=10689&resolved=false&types=BUG) [![Vulnerability](https://camo.githubusercontent.com/3ba1ee49636ffc3427e38649a9f8a65ee392f28e8a662fcf96ce24cefbb520e9/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f76756c6e65726162696c6974792d313670782e706e67)](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=10689&resolved=false&types=VULNERABILITY) [![A](https://camo.githubusercontent.com/1cba125a897d7fa47033a3b3b2be2bbee680d34d4f004a215564659b853fb201/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f412d313670782e706e67)](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=10689&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=10689&resolved=false&types=VULNERABILITY) [![Security Hotspot](https://camo.githubusercontent.com/fb735cbe76f8d5e1679c76ce83b740ceb1eaf62de4f7bf88623dc9953261aff7/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f73656375726974795f686f7473706f742d313670782e706e67)](https://sonarcloud.io/project/security_hotspots?id=apache-dolphinscheduler&pullRequest=10689&resolved=false&types=SECURITY_HOTSPOT) [![A](https://camo.githubusercontent.com/1cba125a897d7fa47033a3b3b2be2bbee680d34d4f004a215564659b853fb201/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f412d313670782e706e67)](https://sonarcloud.io/project/security_hotspots?id=apache-dolphinscheduler&pullRequest=10689&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=apache-dolphinscheduler&pullRequest=10689&resolved=false&types=SECURITY_HOTSPOT) [![Code Smell](https://camo.githubusercontent.com/8fe18b2dfb6f7d4e44582f281b29f617eb5ae07c248d2002ca586e91da219212/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f636f64655f736d656c6c2d313670782e706e67)](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=10689&resolved=false&types=CODE_SMELL) [![A](https://camo.githubusercontent.com/1cba125a897d7fa47033a3b3b2be2bbee680d34d4f004a215564659b853fb201/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f412d313670782e706e67)](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=10689&resolved=false&types=CODE_SMELL) [14 Code Smells](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=10689&resolved=false&types=CODE_SMELL)
> 
> [![0.4%](https://camo.githubusercontent.com/3f04cff3eeef8477afe696ae55c570cbb6ed02f16152497c14251828329a3e91/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f436f76657261676543686172742f302d313670782e706e67)](https://sonarcloud.io/component_measures?id=apache-dolphinscheduler&pullRequest=10689&metric=new_coverage&view=list) [0.4% Coverage](https://sonarcloud.io/component_measures?id=apache-dolphinscheduler&pullRequest=10689&metric=new_coverage&view=list) [![2.2%](https://camo.githubusercontent.com/8047c63e1f9ed03f63001e1eadce4676bade3e0f83ec690a9c625287796248a6/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f4475706c69636174696f6e732f332d313670782e706e67)](https://sonarcloud.io/component_measures?id=apache-dolphinscheduler&pullRequest=10689&metric=new_duplicated_lines_density&view=list) [2.2% Duplication](https://sonarcloud.io/component_measures?id=apache-dolphinscheduler&pullRequest=10689&metric=new_duplicated_lines_density&view=list)

After replacing `@Data`, I think `sonar analysis` will no more fail for low test coverage.

106umao replied:
thx, have done.

106umao commented:
Hi @zhuangchong @SbloodyS @Amy0104 
please help me restart CI failed

SbloodyS commented:
> Hi @zhuangchong @SbloodyS @Amy0104 please help me restart CI failed

Merging the latest dev code can effectively reduce the probability of E2E failure.

106umao commented:
> > Hi @zhuangchong @SbloodyS @Amy0104 please help me restart CI failed
> 
> Merging the latest dev code can effectively reduce the probability of E2E failure.

Thank you for your guidance. Be sure to do so next time.

zhongjiajie commented on docs/docs/en/guide/task/java.md:
```diff
@@ -0,0 +1,48 @@
+- Module Path
+- Main Parameters
+- Java VM Parameters
+
```
```suggestion
```

EricGao888 commented on docs/docs/en/guide/task/java.md:
```diff
@@ -0,0 +1,48 @@
+# Overview
+
+This node is used to perform java-type tasks and supports the use of single files and jar packages as program entries.
```
```suggestion
This node is for executing java-type tasks and supports using files and jar packages as program entries.
```

Please avoid using passive tense in tech docs.

zhongjiajie commented on docs/docs/en/guide/task/java.md:
```diff
@@ -0,0 +1,48 @@
+# Task Parameters
+| **Parameter** | **Description** |
+| ------- | ---------- |
+|Node Name|the name of the set task. The node name in a workflow definition is unique.
+|Run Flag|indicates whether the node is scheduled properly and, if it is not needed, turns on the kill switch.
+|Description|describes the functionality of the node.
+|Task Priority|when the number of worker threads is insufficient, the worker is executed according to the priority from high to low. When the priority is the same, the worker is executed according to the first in, first out principle.
+|Worker Group|The machine whose task is assigned to the Worker group executes, and selecting Default will randomly select a Worker machine to execute.
+|Environment Name|configure the environment in which the task runs.
+|Number Of Failed Retries|number of resubmitted tasks that failed, supported drop-down and hand-fill.
+|Failed Retry Interval|the interval between tasks that fail and are resubmitted, supported by drop-down and hand-fill.
+|Delayed Execution Time|the amount of time a task is delayed, in units.
+|Timeout Alarm|Check timeout warning, timeout failure, when the task exceeds the“Timeout length”, send a warning message and the task execution fails.
+|Module Path|turn on the use of Java 9 + 's modularity feature, put all resources into-module-path, and require that the JDK version in your worker support modularity.
+|Main Parameter|as a normal Java program main method entry parameter.
+|Java VM Parameters|configure startup virtual machine parameters.
+|Script|you need to write Java code if you use the Java run type. The public class must exist in the code without writing a package statement.
+|Resources|these can be external JAR packages or other resource files that are added to the Classpath or module path and can be easily retrieved in your JAVA script.
+|Custom parameter|a user-defined parameter that is part of HTTP and replaces the contents of the script with the ${ variable } .
+|Pre Tasks|selecting a pre-task for the current task sets the selected pre-task upstream of the current task.
```
can you add `|` in the end of this table to make it supports the full Markdown table format? 

```diff
- |Node Name|the name of the set task. The node name in a workflow definition is unique.
+ | Node Name | the name of the set task. The node name in a workflow definition is unique. |
```

106umao replied:
thanks,  have modified.

EricGao888 commented on docs/docs/en/guide/task/java.md:
```diff
@@ -0,0 +1,48 @@
+
+# Create Tasks
+
+- Click on project management-> Project name-> workflow definition, click on the“Create workflow” button, go to the DAG edit page:
```
```suggestion
- Click on `Project Management` -> `Project Name` -> `Workflow Definition`, click on the“Create workflow” button, go to the DAG edit page:
```
Please keep letter case and space usage consistent.

zhongjiajie commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaParameters.java:
```diff
@@ -0,0 +1,89 @@
+    private List<ResourceInfo> resourceList;
+
+    /**
+     * @description:
+     * @date: 7/22/22 2:35 AM
+     * @param: []
```
can we remove this unnecessary part? same as function `getResourceFilesList`

zhongjiajie replied:
I find it still in class `JavaTask.java`, could you search global and remove them?

106umao replied:
> can we remove this unnecessary part? same as function `getResourceFilesList`

this part still left cause extend the AbstractParameter, I not sure can delete

kezhenxu94 replied:
@zhongjiajie might mean to remove the meaningless comments. @106umao  I comment in other places where the comments are meaningless. Please remove them or reword them to more meaningful comments

EricGao888 commented on docs/docs/en/guide/task/java.md:
```diff
@@ -0,0 +1,48 @@
+|Failed Retry Interval|the interval between tasks that fail and are resubmitted, supported by drop-down and hand-fill.
+|Delayed Execution Time|the amount of time a task is delayed, in units.
+|Timeout Alarm|Check timeout warning, timeout failure, when the task exceeds the“Timeout length”, send a warning message and the task execution fails.
+|Module Path|turn on the use of Java 9 + 's modularity feature, put all resources into-module-path, and require that the JDK version in your worker support modularity.
```
```suggestion
|Module Path| pick Java 9 + 's modularity feature, put all resources into-module-path, and require that the JDK version in your worker supports modularity.
```

EricGao888 commented on docs/docs/en/guide/task/java.md:
```diff
@@ -0,0 +1,48 @@
+# Task Parameters
+| **Parameter** | **Description** |
+| ------- | ---------- |
+|Node Name|the name of the set task. The node name in a workflow definition is unique.
```
```suggestion
|Node Name|The name of the set task. The node name in a workflow definition is unique.
```
It doesn't matter you choose to use lower or upper case here, just keep things consistent.

EricGao888 commented on docs/docs/en/guide/task/java.md:
```diff
@@ -0,0 +1,48 @@
+| **Parameter** | **Description** |
+| ------- | ---------- |
+|Node Name|the name of the set task. The node name in a workflow definition is unique.
+|Run Flag|indicates whether the node is scheduled properly and, if it is not needed, turns on the kill switch.
```
```suggestion
|Run Flag|Indicates whether the node is scheduled properly and turns on the kill switch, if not needed.
```
Try to be brief and concise.

EricGao888 commented on docs/docs/en/guide/task/java.md:
```diff
@@ -0,0 +1,48 @@
+| ------- | ---------- |
+|Node Name|the name of the set task. The node name in a workflow definition is unique.
+|Run Flag|indicates whether the node is scheduled properly and, if it is not needed, turns on the kill switch.
+|Description|describes the functionality of the node.
```
```suggestion
|Description|Describes the functionality of the node.
```
As stated above, you could either use lower case or upper case, just be consistent.

zhongjiajie reviewed (commented):
Others LGTM

EricGao888 commented on docs/docs/en/guide/task/java.md:
```diff
@@ -0,0 +1,48 @@
+|Node Name|the name of the set task. The node name in a workflow definition is unique.
+|Run Flag|indicates whether the node is scheduled properly and, if it is not needed, turns on the kill switch.
+|Description|describes the functionality of the node.
+|Task Priority|when the number of worker threads is insufficient, the worker is executed according to the priority from high to low. When the priority is the same, the worker is executed according to the first in, first out principle.
```
```suggestion
|Task Priority|When the number of worker threads is insufficient, the worker executes tasks according to the priority. When the priority is the same, the worker executes tasks by order.

EricGao888 commented on docs/docs/en/guide/task/java.md:
```diff
@@ -0,0 +1,48 @@
+|Run Flag|indicates whether the node is scheduled properly and, if it is not needed, turns on the kill switch.
+|Description|describes the functionality of the node.
+|Task Priority|when the number of worker threads is insufficient, the worker is executed according to the priority from high to low. When the priority is the same, the worker is executed according to the first in, first out principle.
+|Worker Group|The machine whose task is assigned to the Worker group executes, and selecting Default will randomly select a Worker machine to execute.
```
```suggestion
|Worker Group|The group of machines who execute the tasks. If selecting `Default`, DolphinScheduler will randomly choose a worker machine to execute the task.
```

EricGao888 commented on docs/docs/en/guide/task/java.md:
```diff
@@ -0,0 +1,48 @@
+|Description|describes the functionality of the node.
+|Task Priority|when the number of worker threads is insufficient, the worker is executed according to the priority from high to low. When the priority is the same, the worker is executed according to the first in, first out principle.
+|Worker Group|The machine whose task is assigned to the Worker group executes, and selecting Default will randomly select a Worker machine to execute.
+|Environment Name|configure the environment in which the task runs.
```
```suggestion
|Environment Name|Configure the environment in which the task runs.
```

EricGao888 commented on docs/docs/en/guide/task/java.md:
```diff
@@ -0,0 +1,48 @@
+|Task Priority|when the number of worker threads is insufficient, the worker is executed according to the priority from high to low. When the priority is the same, the worker is executed according to the first in, first out principle.
+|Worker Group|The machine whose task is assigned to the Worker group executes, and selecting Default will randomly select a Worker machine to execute.
+|Environment Name|configure the environment in which the task runs.
+|Number Of Failed Retries|number of resubmitted tasks that failed, supported drop-down and hand-fill.
```
```suggestion
|Number Of Failed Retries|Number of resubmitted tasks that failed. You can choose the number in the drop-down menu or fill it manually.
```

EricGao888 commented on docs/docs/en/guide/task/java.md:
```diff
@@ -0,0 +1,48 @@
+|Worker Group|The machine whose task is assigned to the Worker group executes, and selecting Default will randomly select a Worker machine to execute.
+|Environment Name|configure the environment in which the task runs.
+|Number Of Failed Retries|number of resubmitted tasks that failed, supported drop-down and hand-fill.
+|Failed Retry Interval|the interval between tasks that fail and are resubmitted, supported by drop-down and hand-fill.
```
```suggestion
|Failed Retry Interval|the interval between the failure and resubmission of a task. You can choose the number in the drop-down menu or fill it manually.
```

EricGao888 commented on docs/docs/en/guide/task/java.md:
```diff
@@ -0,0 +1,48 @@
+|Delayed Execution Time|the amount of time a task is delayed, in units.
+|Timeout Alarm|Check timeout warning, timeout failure, when the task exceeds the“Timeout length”, send a warning message and the task execution fails.
+|Module Path|turn on the use of Java 9 + 's modularity feature, put all resources into-module-path, and require that the JDK version in your worker support modularity.
+|Main Parameter|as a normal Java program main method entry parameter.
```
```suggestion
|Main Parameter|Java program main method entry parameter.
```

EricGao888 commented on docs/docs/en/guide/task/java.md:
```diff
@@ -0,0 +1,48 @@
+|Timeout Alarm|Check timeout warning, timeout failure, when the task exceeds the“Timeout length”, send a warning message and the task execution fails.
+|Module Path|turn on the use of Java 9 + 's modularity feature, put all resources into-module-path, and require that the JDK version in your worker support modularity.
+|Main Parameter|as a normal Java program main method entry parameter.
+|Java VM Parameters|configure startup virtual machine parameters.
```
```suggestion
|Java VM Parameters|JVM startup parameters.
```

EricGao888 commented on docs/docs/en/guide/task/java.md:
```diff
@@ -0,0 +1,48 @@
+|Module Path|turn on the use of Java 9 + 's modularity feature, put all resources into-module-path, and require that the JDK version in your worker support modularity.
+|Main Parameter|as a normal Java program main method entry parameter.
+|Java VM Parameters|configure startup virtual machine parameters.
+|Script|you need to write Java code if you use the Java run type. The public class must exist in the code without writing a package statement.
```
```suggestion
|Script|You need to write Java code if you use the Java run type. The public class must exist in the code without writing a package statement.
```

EricGao888 commented on docs/docs/en/guide/task/java.md:
```diff
@@ -0,0 +1,48 @@
+|Main Parameter|as a normal Java program main method entry parameter.
+|Java VM Parameters|configure startup virtual machine parameters.
+|Script|you need to write Java code if you use the Java run type. The public class must exist in the code without writing a package statement.
+|Resources|these can be external JAR packages or other resource files that are added to the Classpath or module path and can be easily retrieved in your JAVA script.
```
```suggestion
|Resources|External JAR packages or other resource files that are added to the classpath or module path and can be easily retrieved in your JAVA script.
```

EricGao888 commented on docs/docs/en/guide/task/java.md:
```diff
@@ -0,0 +1,48 @@
+|Java VM Parameters|configure startup virtual machine parameters.
+|Script|you need to write Java code if you use the Java run type. The public class must exist in the code without writing a package statement.
+|Resources|these can be external JAR packages or other resource files that are added to the Classpath or module path and can be easily retrieved in your JAVA script.
+|Custom parameter|a user-defined parameter that is part of HTTP and replaces the contents of the script with the ${ variable } .
```
```suggestion
|Custom parameter|A user-defined parameter that is part of HTTP and replaces `${ variable }` in the script .
```

EricGao888 commented on docs/docs/en/guide/task/java.md:
```diff
@@ -0,0 +1,48 @@
+|Script|you need to write Java code if you use the Java run type. The public class must exist in the code without writing a package statement.
+|Resources|these can be external JAR packages or other resource files that are added to the Classpath or module path and can be easily retrieved in your JAVA script.
+|Custom parameter|a user-defined parameter that is part of HTTP and replaces the contents of the script with the ${ variable } .
+|Pre Tasks|selecting a pre-task for the current task sets the selected pre-task upstream of the current task.
```
```suggestion
|Pre Tasks|Selects a pre-task for the current task and sets the pre-task as the upstream of the current task.
```

EricGao888 commented on docs/docs/en/guide/task/java.md:
```diff
@@ -0,0 +1,48 @@
+
+## Note
+
+When you run a type with JAVA, the public class must exist in the code, and you can not write a package statement.
```
```suggestion
When you run the task in JAVA execution mode, the public class must exist in the code, and you could omit writing a package statement.
```
I'm not quite sure what this sentence means.

EricGao888 commented on docs/docs/en/guide/task/java.md:
```diff
@@ -0,0 +1,48 @@
+
+## Example
+
+Java task types have two modes of execution, which are demonstrated here as an example of a Java mode.
```
```suggestion
Java type tasks have two modes of execution, here is a demonstration of executing tasks in Java mode.
```
I'm not sure what this sentence means. Could you please help make it more comprehensible?

EricGao888 reviewed (changes requested):
I add some comments on the English version docs. Thanks.

106umao committed:
Update docs/docs/en/guide/task/java.md
Apply suggestions from code review

Co-authored-by: Eric Gao <ericgao.apache@gmail.com>
Co-authored-by: Jiajie Zhong <zhongjiajie955@gmail.com>

106umao commented:
> I add some comments on the English version docs. Thanks.

Thanks，I still need to learn from you

106umao committed:
adjust the format of table.

EricGao888 commented:
> > I add some comments on the English version docs. Thanks.
> 
> Thanks，I still need to learn from you

@106umao You're welcome. We are lucky to have a devoted contributor like you. Thanks again for your effort.

EricGao888 commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/exception/JavaSourceFileExistException.java:
```diff
@@ -0,0 +1,39 @@
+
+package org.apache.dolphinscheduler.plugin.task.java.exception;
+
+public class JavaSourceFileExistException extends RuntimeException {
```
It would be better if we could have those exceptions covered in UTs.

EricGao888 reviewed (dismissed):
Backend part overall LGTM, it would be better if we could have those exceptions covered in UTs.

EricGao888 commented:
> Backend part overall LGTM, it would be better if we could have those exceptions covered in UTs.

@SbloodyS Could u please take another look when available? Thanks~

SbloodyS commented:
> > Backend part overall LGTM, it would be better if we could have those exceptions covered in UTs.
> 
> @SbloodyS Could u please take another look when available? Thanks~

Will take a look this afternoon.

106umao committed:
cover exception unit test

106umao commented:
> Backend part overall LGTM, it would be better if we could have those exceptions covered in UTs.

thx, have done

EricGao888 reviewed (dismissed):
Backend part LGTM as long as CI passes, thanks @106umao for the efforts.

106umao commented:
Hi, @SbloodyS PTAL

SbloodyS reviewed (dismissed):
Backend part LGTM. Thanks.

EricGao888 reviewed (dismissed):
Backend LGTM

zhuangchong commented on dolphinscheduler-ui/src/locales/zh_CN/project.ts:
```diff
@@ -284,6 +284,10 @@ export default {
     online: '已上线'
   },
   node: {
+    is_module_path: '模块路劲',
```
Please modify the value.

106umao committed:
optimize locales of project.

SbloodyS reviewed (dismissed):
Backend part LGTM

kezhenxu94 commented on docs/docs/en/guide/task/java.md:
```diff
@@ -0,0 +1,47 @@
+
+# Create Tasks
+
+- Click on `Project Management` -> `Project Name` -> `Workflow Definition`, click on the“Create workflow” button, go to the DAG edit page:
```
```suggestion
- Click on `Project Management` -> `Project Name` -> `Workflow Definition`, click on the “Create workflow” button, go to the DAG edit page:
```

kezhenxu94 commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaConstants.java:
```diff
@@ -0,0 +1,67 @@
+        throw new IllegalStateException("Utility class");
+    }
+
+    /**
+     * constant ${JAVA_HOME}
+     **/
```
All these comments in this file is meaningless

kezhenxu94 commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaParameters.java:
```diff
@@ -0,0 +1,89 @@
+     */
+    private ResourceInfo mainJar;
+
+    /**
+     * run type is JAVA or JAR
+     */
+    private String runType;
```
Consider making it an enum type

kezhenxu94 commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaTask.java:
```diff
@@ -0,0 +1,408 @@
+     */
+    private JavaParameters javaParameters;
+
+    /**
+     * shell command executor
+     */
```
Please remove all these useless comments

kezhenxu94 commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaTask.java:
```diff
@@ -0,0 +1,408 @@
+            }
+            Preconditions.checkNotNull(command, "command not be null.");
+            TaskResponse taskResponse = shellCommandExecutor.run(command);
+            logger.info("java task run result : " + taskResponse);
```
```suggestion
            logger.info("java task run result: {}", taskResponse);
```

106umao committed:
fix typo of the java.md and optimize the code comment of java file.

106umao committed:
optimize log output.

EricGao888 commented:
I've restarted the failed CI.

zhuangchong commented:
@SbloodyS @zhongjiajie @EricGao888 Please help with code review, thanks.

zhongjiajie commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaParameters.java:
```diff
@@ -0,0 +1,89 @@
+
+    /**
+     * @description: Check that the parameters are valid
+     * @date: 7/22/22 2:35 AM
```
I think we should remove the `date` comment for all new adding or change files

zhongjiajie commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaTask.java:
```diff
@@ -0,0 +1,438 @@
+        return mainJar.getId() == 0
+                ? mainJar.getRes()
+                // when update resource maybe has error
+                : mainJar.getResourceName().replaceFirst("/", "");
```
we have slash in class constant, so as
* `PERIOD` for `"."`
* `SPACE` for `" "`
```suggestion
                : mainJar.getResourceName().replaceFirst(Constants.SINGLE_SLASH, "");
```

zhongjiajie reviewed (commented):
thanks for adding this, others LGTM

SbloodyS commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaTask.java:
```diff
@@ -0,0 +1,438 @@
+     * @return: java.lang.String
+     **/
+    protected void createJavaSourceFileIfNotExists(String sourceCode, String fileName) throws IOException {
+        logger.info("tenantCode :{}, task dir:{}", taskRequest.getTenantCode(), taskRequest.getExecutePath());
```
```suggestion
        logger.info("tenantCode: {}, task dir: {}", taskRequest.getTenantCode(), taskRequest.getExecutePath());
```

SbloodyS commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaTask.java:
```diff
@@ -0,0 +1,438 @@
+        logger.info("tenantCode :{}, task dir:{}", taskRequest.getTenantCode(), taskRequest.getExecutePath());
+
+        if (!Files.exists(Paths.get(fileName))) {
+            logger.info("generate java source file:{}", fileName);
```
```suggestion
            logger.info("generate java source file: {}", fileName);
```

SbloodyS commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaTask.java:
```diff
@@ -0,0 +1,438 @@
+        shellCommandExecutor = new ShellCommandExecutor(this::logHandle,
+                taskRequest,
+                logger);
+        logger.info("java task code compile result : " + compileResponse);
```
```suggestion
        logger.info("java task code compile result: " + compileResponse);
```

SbloodyS commented on dolphinscheduler-ui/src/locales/en_US/project.ts:
```diff
@@ -143,7 +142,7 @@ export default {
     stop: 'Stop',
     pause: 'Pause',
-    recovery_waiting_thread: 'Recovery waiting thread',
+    recovery_waiting_thread:  'Recovery waiting thread',
```
Please remove unnessnary change.

EricGao888 commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaParameters.java:
```diff
@@ -0,0 +1,89 @@
+import lombok.Setter;
+import lombok.ToString;
+
+@Getter
```
Sorry for the trouble, could you please use `@Data` here? We have fixed the `sonar` to exclude generated code and we'd better keep things consistent as discussed in https://github.com/apache/dolphinscheduler/issues/11074
Thanks!

106umao replied:
done

SbloodyS commented on dolphinscheduler-ui/src/locales/zh_CN/project.ts:
```diff
@@ -284,6 +284,10 @@ export default {
     online: '已上线'
   },
   node: {
+    is_module_path: '使用模块路劲',
```
```suggestion
    is_module_path: '使用模块路径',
```

106umao committed:
change lombok, remove data comment.

optimize log and constant

EricGao888 reviewed (dismissed):
LGTM as long as CI passes.

SbloodyS commented on docs/docs/zh/guide/task/java.md:
```diff
@@ -0,0 +1,50 @@
+- 失败重试间隔：任务失败重新提交任务的时间间隔，支持下拉和手填。
+- 延迟执行时间：任务延迟执行的时间，以分为单位。
+- 超时告警：勾选超时告警、超时失败，当任务超过"超时时长"后，会发送告警邮件并且任务执行失败。
+- 模块路劲：开启使用JAVA9+的模块化特性，把所有资源放入--module-path中，要求您的worker中的JDK版本支持模块化。
```
```suggestion
- 模块路径：开启使用JAVA9+的模块化特性，把所有资源放入--module-path中，要求您的worker中的JDK版本支持模块化。
```

106umao replied:
thanks

106umao commented:
> LGTM as long as CI passes.

thanks,  I have done some work again. PTAL

SbloodyS reviewed (dismissed):
The backend parts LGTM. Thanks.

106umao commented:
somebody can help me look on the docs dead link error:  [✖] https://dolphinscheduler.apache.org/python/index.html → Status: 404

106umao committed:
fix build failed

optimize

zhuangchong commented:
> somebody can help me look on the docs dead link error: [✖] https://dolphinscheduler.apache.org/python/index.html → Status: 404

Please merge the latest dev branch code.

106umao commented:
> > somebody can help me look on the docs dead link error: [heavy_multiplication_x] https://dolphinscheduler.apache.org/python/index.html → Status: 404
> 
> Please merge the latest dev branch code.

have done, pls reaview for me

ruanwenjun commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaConstants.java:
```diff
@@ -0,0 +1,67 @@
+    /**
+     * This constant is the Classpath or module path delimiter for different operating systems
+     **/
+    public static final String PATH_SEPARATOR = System.getProperty("path.separator");
```
Why you don't directly use `File.pathSeparatorChar`

ruanwenjun commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaConstants.java:
```diff
@@ -0,0 +1,67 @@
+    /**
+     * This constant is the file delimiter for different operating systems
+     **/
+    public static final String FILE_SEPARATOR = System.getProperty("file.separator");
```
Why you don't directly use `File.separator`

106umao replied:
thx

ruanwenjun commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaParameters.java:
```diff
@@ -0,0 +1,83 @@
+     **/
+    @Override
+    public boolean checkParameters() {
+        return runType != null && (rawScript != null && !rawScript.isEmpty()) || mainJar != null;
```
```suggestion
        return runType != null && StringUtils.isNotEmpty(rawScript) || mainJar != null;
```

ruanwenjun commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaTask.java:
```diff
@@ -0,0 +1,422 @@
+
+import com.google.common.base.Preconditions;
+
+/**
+ * java task
+ */
```
```suggestion

```

ruanwenjun commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaTask.java:
```diff
@@ -0,0 +1,422 @@
+     */
+    private static final Pattern classNamePattern = Pattern.compile(PUBLIC_CLASS_NAME_REGEX);
+
+    /**
+     * @description: Construct a Java task
+     * @param: [org.apache.dolphinscheduler.plugin.task.api.TaskExecutionContext]
+     * @return: JavaTask
+     **/
```
This kind of comment is not needed.

106umao replied:
sure

ruanwenjun commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaTask.java:
```diff
@@ -0,0 +1,422 @@
+        if (javaParameters == null || !javaParameters.checkParameters()) {
+            throw new TaskException("java task params is not valid");
+        }
+        if (javaParameters.getRunType().equals(JavaConstants.RUN_TYPE_JAR)) {
```
```suggestion
        if (JavaConstants.RUN_TYPE_JAR.equals(javaParameters.getRunType())) {
```

ruanwenjun commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaTask.java:
```diff
@@ -0,0 +1,422 @@
+        try {
+            rawJavaScript = convertJavaSourceCodePlaceholders(rawJavaScript);
+        } catch (StringIndexOutOfBoundsException e) {
+            logger.error("setShareVar field format error, raw java script : {}", rawJavaScript);
```
```suggestion
            logger.error("setShareVar field format error, raw java script : {}", rawJavaScript, e);
```

ruanwenjun commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaTask.java:
```diff
@@ -0,0 +1,422 @@
+                case JavaConstants.RUN_TYPE_JAR:
+                    command = buildJarCommand();
+                    break;
+                default:
+                    throw new RunTypeNotFoundException("run type is required, but it is null now.");
```
Why you don't put this in `javaParameters.checkParameters()`

ruanwenjun commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaTask.java:
```diff
@@ -0,0 +1,422 @@
+     * @param: []
+     * @return: java.lang.String
+     **/
+    protected String buildJavaCommand() throws Exception {
```
```suggestion
    protected @Nonnull String buildJavaCommand() throws Exception {
```

ruanwenjun commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaTask.java:
```diff
@@ -0,0 +1,422 @@
+            logger.error("java task interrupted ", e);
+            setExitStatusCode(TaskConstants.EXIT_CODE_FAILURE);
+            Thread.currentThread().interrupt();
+        } catch (RunTypeNotFoundException e) {
+            logger.error(e.getMessage());
+            setExitStatusCode(TaskConstants.EXIT_CODE_FAILURE);
+            throw e;
```
```suggestion
        
```
This catch is not needed, if you want to print the exception, it's better to print the thread stack rather than e.getMessage.

ruanwenjun commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaTask.java:
```diff
@@ -0,0 +1,422 @@
+    }
+
+    /**
+     * @description: Construct a shell command for the java -jar Run mode
+     * @param: []
+     * @return: java.lang.String
+     **/
```
```suggestion
     * @description: Construct a shell command for the java -jar Run mode
     * @return: java.lang.String
     **/
```

ruanwenjun commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaTask.java:
```diff
@@ -0,0 +1,422 @@
+    }
+
+    /**
+     * @description: Construct a shell command for the java Run mode
+     * @param: []
+     * @return: java.lang.String
+     **/
```
```suggestion
     * @description: Construct a shell command for the java Run mode
     * @return: java.lang.String
     **/
```

ruanwenjun commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaTask.java:
```diff
@@ -0,0 +1,422 @@
+    }
+
+    /**
+     * @description: Replaces placeholders such as local variables in source files
+     * @param: [java.lang.String]
+     * @return: java.lang.String
```
AFAIK, this is not the standard javadoc?

ruanwenjun commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaTask.java:
```diff
@@ -0,0 +1,422 @@
+     * @param: [java.lang.String]
+     * @return: java.lang.String
+     **/
+    protected static String convertJavaSourceCodePlaceholders(String rawScript) throws StringIndexOutOfBoundsException {
```
Can we directly use `ParameterUtils.convertParameterPlaceholders`?

106umao replied:
Maybe. I haven't tried

106umao replied:
This method has the wrong comment. It's not the same method as what you said. Sorry

ruanwenjun commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaTask.java:
```diff
@@ -0,0 +1,422 @@
+        if (!Files.exists(Paths.get(fileName))) {
+            logger.info("generate java source file: {}", fileName);
+
+            StringBuilder sb = new StringBuilder();
+            sb.append(sourceCode);
+            logger.info(sb.toString());
+
+            // write data to file
+            FileUtils.writeStringToFile(new File(fileName),
+                    sb.toString(),
+                    StandardCharsets.UTF_8);
```
```suggestion
            logger.info("The java source code", sourceCode);

            // write data to file
            FileUtils.writeStringToFile(new File(fileName),
                    sb,
                    StandardCharsets.UTF_8);
```
Why do you use StringBuilder here?

106umao replied:
Thank you. The test code is not clean

ruanwenjun commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaTask.java:
```diff
@@ -0,0 +1,422 @@
+        Preconditions.checkNotNull(compileCommand, "command not be null.");
+        TaskResponse compileResponse = shellCommandExecutor.run(compileCommand);
+        // must drop the command file ,if do not ,the next command will not run. because be limited the ShellCommandExecutor's create file rules
+        dropShellCommandFile();
```
This should be a common logic for shell task?

106umao replied:
thx, It is a bit strange to have two processes running, I will optimize to run in one process

Amy0104 commented on dolphinscheduler-ui/src/views/projects/task/components/node/fields/use-java.ts:
```diff
@@ -0,0 +1,86 @@
+      validate: {
+        trigger: ['input', 'trigger'],
+        required: true,
+        message: t('project.node.script_tips')
```
Does it have a language? The default language of the editor is shell.

106umao replied:
Usually no one writes Java code in a browser, so a shell should suffice

Amy0104 commented on dolphinscheduler-ui/src/views/projects/task/constants/task-type.ts:
```diff
@@ -47,6 +48,9 @@ export type TaskType =
 export const TASK_TYPES_MAP = {
+  JAVA: {
+    alias: 'JAVA'
+  },
```
Does it have a document, if not, you need to set helperLinkDisable to true.

106umao replied:
Yes

106umao committed:
resolve conflict

106umao committed:
Update dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaParameters.java

Co-authored-by: Wenjun Ruan <wenjun@apache.org>

zhuangchong commented:
Please solve the ci check license error.

ruanwenjun commented on pom.xml:
```diff
@@ -261,6 +262,12 @@
+                <artifactId>mysql-connector-java</artifactId>
+                <version>${mysql.connector.version}</version>
+            </dependency>
+
```
Please remove this dependency, we should already define this in dolphinscheduler-bom

106umao replied:
have done

ruanwenjun reviewed (dismissed):
LGTM.

ruanwenjun commented on dolphinscheduler-task-plugin/dolphinscheduler-task-java/src/main/java/org/apache/dolphinscheduler/plugin/task/java/JavaParameters.java:
```diff
@@ -68,7 +69,7 @@ public class JavaParameters extends AbstractParameters {
     @Override
     public boolean checkParameters() {
-        return runType != null && (rawScript != null && !rawScript.isEmpty()) || mainJar != null;
+        return runType != null && StringUtils.isNotBlank(rawScript) || mainJar != null;
```
```suggestion
        return runType != null && (StringUtils.isNotBlank(rawScript) || mainJar != null);
```

When the runType is jar you need to check if the `mainJar != null`, when the runType is script you need to check if the `rawScript != null`

106umao committed:
resolve conflict

zhuangchong reviewed (approved):
LGTM.

EricGao888 commented:
@106umao Congrats! Thanks for the contributions : )
